### PR TITLE
Fix LLVM error

### PR DESCRIPTION
**Pull request by orbea**
This will fix the following crash when trying to start dolphin-emu with llvm on some distros.
```
: CommandLine Error: Option 'aarch64-neon-syntax' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
```
The fix is courtesy of @Orphis and this started in the following commit.
```
commit f649e26655891411df99c072344d7bc4a9c25db6
Author: Florent Castelli <florent.castelli@gmail.com>
Date:   Wed Feb 8 02:59:41 2017 +0100

    cmake: Move LLVM import to UICommon
    
    To use it, with a modern LLVM (3.9+), set your CMAKE_PREFIX_PATH
    to point to the LLVM install folder or to a LLVM build folder.
    We're linking ALL of LLVM libs since I don't really know which ones we need.
    LTO will take care of sliming the binary size...

:100644 100644 1eee549f223acf37bff953fe9283dfa5cb98a7b1 f418cf106056e1f4f064b4ccce4c34eada620a61 M	CMakeLists.txt
:040000 040000 fcb9284f7af823bf4e1ee966c742c778e8a1fd44 579782d2d2cfa3852889541e3c031427e4fad075 M	CMakeTests
:040000 040000 b9b264a93531e2a8d8ab1ed5fd347b40fb1df6ba ac2936891a4eba9bbc3278629b1c7ead6b54d8d7 M	Source
```
f649e26655891411df99c072344d7bc4a9c25db6

### CODE REVIEW

**Comment by ligfx** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/UICommon/CMakeLists.txt [hunk 1]
```diff
@@ -14,7 +14,7 @@ if(ENABLE_LLVM)
   if(LLVM_FOUND)
     message(STATUS "LLVM found, enabling LLVM support in disassembler")
     target_compile_definitions(uicommon PRIVATE HAVE_LLVM)
-    target_link_libraries(uicommon PRIVATE ${LLVM_AVAILABLE_LIBS})
+    target_link_libraries(uicommon PRIVATE LLVM)
```
Comment (endrift) [inline Source/Core/UICommon/CMakeLists.txt]: Should be `${LLVM_LIBRARIES}`

Comment (endrift) [inline Source/Core/UICommon/CMakeLists.txt]: Scratch that, apparently the LLVMConfig thing is unconventional. Regardless, this change is wrong.

**Comment by orbea** [comment]
@Orphis Can you suggest a better fix that would satisfy the complaints?

**Comment by endrift** [comment]
I might try seeing if this passes buildbot. It's a bit of a hack though.

```patch
diff --git a/Source/Core/UICommon/CMakeLists.txt b/Source/Core/UICommon/CMakeLists.txt
index b76e79a..11bcb30 100644
--- a/Source/Core/UICommon/CMakeLists.txt
+++ b/Source/Core/UICommon/CMakeLists.txt
@@ -12,9 +12,15 @@ add_dolphin_library(uicommon "${SRCS}" "${LIBS}")
 if(ENABLE_LLVM)
   find_package(LLVM CONFIG QUIET)
   if(LLVM_FOUND)
+    set(LLVM_REGEX "\\bLLVM(-3\\.\\d+)?\\b")
+    if("${LLVM_AVAILABLE_LIBS}" MATCHES "${LLVM_REGEX}")
+      string(REGEX MATCH LLVM_LIB "${LLVM_REGEX}" "${LLVM_AVAILABLE_LIBS}")
+    else()
+      set(LLVM_LIB "${LLVM_AVAILABLE_LIBS}")
+    endif()
     message(STATUS "LLVM found, enabling LLVM support in disassembler")
     target_compile_definitions(uicommon PRIVATE HAVE_LLVM)
-    target_link_libraries(uicommon PRIVATE ${LLVM_AVAILABLE_LIBS})
+    target_link_libraries(uicommon PRIVATE ${LLVM_LIB})
     target_include_directories(uicommon PRIVATE ${LLVM_INCLUDE_DIRS})
   endif()
 endif()
```

**Comment by Orphis** [comment]
It's not a hack and this is the preferred CMake way for library authors to export them to CMake (to replace Find* modules). 

In LLVM's case, they create some imported targets, one of which is called LLVM. That's all. Not a variable.

**Comment by endrift** [comment]
Except it's demonstrably false that LLVMConfig always creates a target called LLVM; it does no such thing with the default FreeBSD LLVM, as FreeBSD does not ship a default libLLVM.so; only the static libraries. The problem with @orbea was that it was linking the static AND shared object. On FreeBSD, it only grabbed the static libraries. This tells it to explicitly look for a thing that the LLVMConfig never explicitly said it exported (or exported at all).

**Comment by Orphis** [comment]
It is more likely that the distribution on FreeBSD and the available targets are different, it doesn't make it invalid. 
Also, LLVM on FreeBSD defaults to 3.7 when he used 3.8, that could explain some differences. I already noticed differences when I made the initial patch between Ubuntu, macOS and FreeBSD, so I wouldn't be surprised.

**Comment by endrift** [comment]
It makes the *approach* invalid, as it's no longer cross-platform to all LLVM distributions. Yes, sure it may be an issue with the LLVM distribution in FreeBSD, but it is still a platform that is (vaguely) supported. The only real solutions are to hack it so it works on FreeBSD (like I did), or just disable in on FreeBSD (also kind of hack).

**Comment by Orphis** [comment]
You can just change it to: 

> if(LLVM_FOUND AND TARGET LLVM)

That will filter out bad distributions of LLVM that don't have libLLVM.so.

**Comment by Orphis** [comment]
Alternatively, we could change the detection to this:
`find_package(LLVM 3.9 CONFIG QUIET)`

**Comment by Orphis** [comment]
After much thought and investigation, I believe checking for the LLVM target directly is the better choice.
LLVM 3.7 and 3.8 on FreeBSD don't build the dynamic libraries, but 3.9 does. So you could build against that if you wanted a more advanced debugger there.
Then, we probably have approximately a grand total of zero (or maybe only you @endrift ) users of the debugger on that platform, so it's not inconceivable to put that requirement.

**Comment by orbea** [comment]
I just want to add that Slackware current upgraded to llvm-3.9.1, and the initial problem remains.

**Comment by orbea** [comment]
@endrift I updated the fix with one of the compromises suggested above, assuming this passes the buildbot would it be satisfactory for you?

**Comment by endrift** [comment]
@orbea I think the target_link_libraries param could be `LLVM` if `LLVM_TARGET`, or `${LLVM_AVAILABLE_LIBS}` otherwise and it would work. Not sure though.

**Comment by Orphis** [comment]
@endrift No, this is fine. The check above "TARGET LLVM" checks there's an actual target called LLVM. If there's one, it comes from the LLVMConfig file and it's ok to link against it.

**Comment by Orphis** [comment]
@dolphin-emu-bot rebuild

**Comment by endrift** [comment]
@Orphis that's not what I meant. I mean that if TARGET_LLVM does not exist, LLVM_AVAILABLE_LIBS will be populated with the static libs, so it should still work in that case.

**Comment by Orphis** [comment]
@endrift Maybe yes, maybe not. I think this is fine for now. If people only who really need the feature only have the static libs and really can't get the shared library on their platform, we may reconsider this choice later. Most likely, they didn't link against LLVM in the previous version at all, so it's not likely to be an issue.

**Comment by orbea** [comment]
@endrift I tend to agree that if it can work for freebsd there should be no reason to disable it.

How about this?
```
diff --git a/Source/Core/UICommon/CMakeLists.txt b/Source/Core/UICommon/CMakeLists.txt
index b76e79a66f..7b9d93a473 100644
--- a/Source/Core/UICommon/CMakeLists.txt
+++ b/Source/Core/UICommon/CMakeLists.txt
@@ -14,7 +14,11 @@ if(ENABLE_LLVM)
   if(LLVM_FOUND)
     message(STATUS "LLVM found, enabling LLVM support in disassembler")
     target_compile_definitions(uicommon PRIVATE HAVE_LLVM)
-    target_link_libraries(uicommon PRIVATE ${LLVM_AVAILABLE_LIBS})
+    if(TARGET LLVM)
+      target_link_libraries(uicommon PRIVATE LLVM)
+    else()
+      target_link_libraries(uicommon PRIVATE ${LLVM_AVAILABLE_LIBS})
+    endif()
     target_include_directories(uicommon PRIVATE ${LLVM_INCLUDE_DIRS})
   endif()
 endif()
```
It works here, but I can't test many other platforms.

**Comment by Orphis** [comment]
No. It works fine on FreeBSD as long as you use a proper LLVM distribution (such as the 3.9 one). No reason to add any hack in there.

**Pull request merged**